go/roothash/api/api.go

```diff
@@ -142,6 +156,150 @@ func NewRequestProposerTimeoutTx(nonce uint64, fee *transaction.Fee, runtimeID c
 	})
 }
 
+// EvidenceKind is the evidence kind.
+type EvidenceKind uint8
+
+const (
+	// EvidenceKindEquivocation is the evidence kind for equivocation.
+	EvidenceKindEquivocation = 1
+)
+
+// Evidence is an evidence of node misbehaviour.
+type Evidence struct {
+	ID common.Namespace `json:"id"`
+
+	EquivocationExecutor *EquivocationExecutorEvidence `json:"equivocation_executor,omitempty"`
+	EquivocationBatch    *EquivocationBatchEvidence    `json:"equivocation_batch,omitempty"`
+}
+
+// Hash computes the evidence hash.
+//
+// Hash is derived by hashing the evidence kind and the public key of the signer.
+// Assumes evidence has been validated.
+func (ev *Evidence) Hash() (hash.Hash, error) {
+	switch {
+	case ev.EquivocationBatch != nil:
+		return hash.NewFromBytes([]byte{EvidenceKindEquivocation}, ev.EquivocationBatch.BatchA.Signature.PublicKey[:]), nil
+	case ev.EquivocationExecutor != nil:
+		return hash.NewFromBytes([]byte{EvidenceKindEquivocation}, ev.EquivocationExecutor.CommitA.Signature.PublicKey[:]), nil
+	default:
+		return hash.Hash{}, fmt.Errorf("cannot compute hash, invalid evidence")
+	}
+}
+
+// ValidateBasic performs basic evidence validity checks.
+func (ev *Evidence) ValidateBasic() error {
+	switch {
+	case ev.EquivocationExecutor != nil && ev.EquivocationBatch != nil:
+		return fmt.Errorf("evidence has multiple fields set")
+	case ev.EquivocationExecutor != nil:
+		return ev.EquivocationExecutor.ValidateBasic()
+	case ev.EquivocationBatch != nil:
+		return ev.EquivocationBatch.ValidateBasic()
+	default:
+		return fmt.Errorf("evidence content has no fields set")
+	}
+}
+
+// EquivocationExecutorEvidence is evidence of executor commitment equivocation.
+type EquivocationExecutorEvidence struct {
+	CommitA commitment.ExecutorCommitment `json:"commit_a"`
+	CommitB commitment.ExecutorCommitment `json:"commit_b"`
+}
+
+// ValidateBasic performs basic executor evidence validation checks.
+// TODO: maybe rename to: Validate(), might better indicate that this actually
+// checks if evidence is valid and is not just a basic check.
+func (ev *EquivocationExecutorEvidence) ValidateBasic() error {
+	a, err := ev.CommitA.Open()
+	if err != nil {
+		return fmt.Errorf("opening CommitA: %w", err)
+	}
+	b, err := ev.CommitB.Open()
+	if err != nil {
+		return fmt.Errorf("opening CommitB: %w", err)
+	}
+
+	if a.Body == nil {
+		return fmt.Errorf("CommitA: body empty")
```

Comment:
is this possible? It looks like not. but if it becomes possible, which is more likely, that it's the evidence submitter's fault or that it's the signer's fault? here we return an error, which otherwise seems to tell the caller something happened that was the submitter's fault

Reply:
If `Body` was `nil` this would actually indicate a bug in the `Open` logic. So this check indeed does not make sense and I'll remove it.